perf/pgbench.py

```diff
@@ -0,0 +1,143 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2023 IBM
+# Author: Gautam Menghani <gautam@linux.ibm.com>
+
+import os
+
+from avocado import Test
+from avocado.utils import process, cpu
+from avocado.utils.service import ServiceManager
+from avocado.utils.software_manager.manager import SoftwareManager
+
+
+class Pgbench(Test):
+
+    """
+    This module will run the pgbench benchmark. Pgbench is a benchmark for
+    measuring the throughput achieved by executing database transactions on
+    PostgresSQL database.
+    """
+    sm = SoftwareManager()
+    scaling_factor = worker_threads = db_clients = protocol = \
+        benchmark_duration = transaction_count = 0
+    already_installed = True
+
+    def run_cmd(self, cmd, ignore_failure):
+        command = process.SubProcess(cmd)
+        command.run()
+
+        if command.result.exit_status and ignore_failure:
+            self.log.info(f"Command '{cmd}' returned '{command.result.exit_status}' "
+                          f"with the text '{command.result.stderr_text}'")
+            return [command.result.exit_status, command.result.stderr_text]
+        elif command.result.exit_status:
+            self.cancel(f"Command '{cmd}' returned '{command.result.exit_status}' "
+                        f"with the text '{command.result.stderr_text}'")
+        else:
+            return [command.result.exit_status, command.result.stdout_text]
+
+    def get_pkg_manager(self):
+        for pkg_mgr in ['apt', 'dnf', 'yum']:
+            if self.sm.check_installed(pkg_mgr):
+                return pkg_mgr
+        return None
+
+    def install_and_initialize_postgres(self, pkg_mgr):
+        packages = ["postgresql", "postgresql-contrib"]
+        for package in packages:
+            self.sm.install(package)
+
+        # rhel needs an extra pkg and initialization
+        if pkg_mgr in ['dnf', 'yum']:
+            self.sm.install('postgresql-server')
+            output = self.run_cmd("postgresql-setup --initdb", True)
+            if output[0] and "not empty" not in output[1]:
+                self.cancel(f"Failed to initialize postgres - {output[1]}")
+
+    def setUp(self):
+        '''
+        Install Pgbench
+        '''
+        # check root privileges ($USER should be able to run commands
+        # as another user without requiring a password)
+        if os.geteuid() != 0:
+            self.cancel("This script requires root privileges, "
+                        "Please try again with 'sudo' or as 'root'")
+
+        # check if pgbench is installed
+        if (self.sm.check_installed("pgbench")):
+            self.log.info("Pgbench is already installed")
+        else:
+            pkg_mgr = self.get_pkg_manager()
+            if pkg_mgr:
+                self.install_and_initialize_postgres(pkg_mgr)
+                self.already_installed = False
+            else:
+                self.cancel("Unsupported Linux distribution")
```

Comment:
instead of the package manager, we should be having checks based on the distro name

Reply:
we used package manger so that we don't have to limit the script to a few distros like debian/ubuntu. If the user is using a debian derivative like ubuntu/pop os/linux mint, we don't have to hardcode all those distros as the package manager will be the same; so we cover more ground if we detect package manager

Reply:
as per abdul's comment above, I am adding support for suse as well

Reply:
sure , some how not liking taking decision based on package manger

Reply:
ok can this break in any scenario? Because if we use distro names and just specify "debian,ubuntu,rhel,fedora,suse",  a user running this on a debian derivative will get "unsupported distro" even if the benchmark can run just fine.

Reply:
or is it that we want to implicitly imply that we have tested this to work on these 4-5 distros? Please advise on this

Reply:
@gautammenghani ideally where we tested we should have supported if any user is looking supported in the next distro we can welcome a patch